Mark PVCDataSource featuregate as GA

j-griffith opened the pull request:
/kind feature


**What this PR does / why we need it**:

Marks VolumePVCDataSource as GA for 1.18

**Which issue(s) this PR fixes**:

Fixes #88814

**Special notes for your reviewer**:

```release-note
VolumePVCDataSource moves to GA in 1.18 release
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:


```docs
https://github.com/kubernetes/website/pull/19318
```

jsafrane commented on pkg/api/persistentvolumeclaim/util_test.go:
```diff
@@ -210,90 +210,3 @@ func TestDropDisabledSnapshotDataSource(t *testing.T) {
-			gateEnabled: false,
-			want:        nil,
-		},
-		"enabled with valid spec but nil APIGroup": {
```
IMO all "enabled" tests are still valid and should be preserved. The best would be to remove `gateEnabled bool` and leave the tests that have it as `true`.

j-griffith replied:
Oops!  That was a mistake on my side; I'll get those back in place.  Thanks for catching this.

jsafrane reviewed (commented):
Please remove feature gate evaluation in https://github.com/kubernetes/kubernetes/blob/fdb2cb4c8832da1499069bda918c014762d8ac05/pkg/api/persistentvolumeclaim/util.go#L67 - it's always `true` now.

jsafrane commented:
/lgtm
Failed e2e-gce-alpha-features is not related to this PR and is being fixed in https://github.com/kubernetes/kubernetes/pull/88727

jsafrane commented:
/kind api-review
for changes in validation (i.e. removal of feature checks)

jsafrane commented:
/label api-review
for changes in validation (i.e. removal of feature checks)

smarterclayton commented:
/assign @smarterclayton

smarterclayton commented:
Does the godoc on `DataSource` need to be updated?  I see:

```
	// This field requires the VolumeSnapshotDataSource alpha feature gate to be
	// enabled and currently VolumeSnapshot is the only supported data source.
	// If the provisioner can support VolumeSnapshot data source, it will create
	// a new volume and data will be restored to the volume at the same time.
	// If the provisioner does not support VolumeSnapshot data source, volume will
	// not be created and the failure will be reported as an event.
	// In the future, we plan to support more data source types and the behavior
	// of the provisioner may change.
	// +optional
	DataSource *TypedLocalObjectReference `json:"dataSource,omitempty" protobuf:"bytes,7,opt,name=dataSource"`
```

and I would expect it to be updated.

j-griffith commented:
> Does the godoc on `DataSource` need to be updated? I see:
> 
> ```
> 	// This field requires the VolumeSnapshotDataSource alpha feature gate to be
> 	// enabled and currently VolumeSnapshot is the only supported data source.
> 	// If the provisioner can support VolumeSnapshot data source, it will create
> 	// a new volume and data will be restored to the volume at the same time.
> 	// If the provisioner does not support VolumeSnapshot data source, volume will
> 	// not be created and the failure will be reported as an event.
> 	// In the future, we plan to support more data source types and the behavior
> 	// of the provisioner may change.
> 	// +optional
> 	DataSource *TypedLocalObjectReference `json:"dataSource,omitempty" protobuf:"bytes,7,opt,name=dataSource"`
> ```
> 
> and I would expect it to be updated.

@smarterclayton You're right, looks like that has been missed the last couple of releases, I've updated; PTAL.

jsafrane commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -463,14 +463,13 @@ type PersistentVolumeClaimSpec struct {
-	// This field requires the VolumeSnapshotDataSource alpha feature gate to be
-	// enabled and currently VolumeSnapshot is the only supported data source.
-	// If the provisioner can support VolumeSnapshot data source, it will create
+	// The Object Type in this field currently can be either a VolumeSnapshot, or an existing CSI PVC.
```
@j-griffith, I think it should be in sync with this: https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L422-L433

They should be the same and not mention `VolumePVCDataSource` feature gate.

In addition, `make update` will update all the generated APIs.

j-griffith replied:
@jsafrane Thanks... I thought that the changes in staging automatically propagated for some reason (or even vice versa) but apparently not, sorry.  So I guess update both?

Commit:
Update VolumePVCDatasource to GA for 1.18

Updates the VolumePVCDataSource featuregate (cloning) to GA for the 1.18 k8s
release.

j-griffith commented:
@smarterclayton updated the docstrings and ran `make generate`, please take a look, thanks!

j-griffith commented:
/test pull-kubernetes-node-e2e-containerd

jsafrane commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -463,14 +463,18 @@ type PersistentVolumeClaimSpec struct {
 	// not be created and the failure will be reported as an event.
 	// In the future, we plan to support more data source types and the behavior
 	// of the provisioner may change.
+
```
Please remove this blank line

jsafrane commented on staging/src/k8s.io/api/core/v1/generated.proto:
```diff
@@ -2640,14 +2640,6 @@ message PersistentVolumeClaimSpec {
-  // not be created and the failure will be reported as an event.
-  // In the future, we plan to support more data source types and the behavior
-  // of the provisioner may change.
   // +optional
```
This is suspicious, probably caused by the blank line below.

jsafrane commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -192,6 +192,7 @@ func NewCmdApply(baseName string, f cmdutil.Factory, ioStreams genericclioptions
 	cmd.Flags().BoolVar(&o.OpenAPIPatch, "openapi-patch", o.OpenAPIPatch, "If true, use openapi to calculate diff when the openapi presents and the resource can be found in the openapi spec. Otherwise, fall back to use baked-in types.")
 	cmd.Flags().Bool("server-dry-run", false, "If true, request will be sent to server with dry-run flag, which means the modifications won't be persisted.")
 	cmd.Flags().MarkDeprecated("server-dry-run", "--server-dry-run is deprecated and can be replaced with --dry-run=server.")
+	cmd.Flags().MarkHidden("server-dry-run")
```
This is suspicious... Is it really part of this PR?

jsafrane reviewed (commented):
Can you please fix the commit message?

j-griffith commented:
/retest

j-griffith commented:
/test pull-kubernetes-e2e-kind

j-griffith commented:
/test pull-kubernetes-verify

j-griffith commented:
/test pull-kubernetes-e2e-gci-gce-autoscaling

j-griffith commented:
/test pull-kubernetes-e2e-gci-gce-autoscaling

j-griffith commented:
/test pull-kubernetes-verify

jsafrane commented:
almost there, but generated doc is out of sync:

```
Generated swagger type documentation is out of date:
--- /home/prow/go/src/k8s.io/kubernetes/./staging/src/k8s.io/api/core/v1/types_swagger_doc_generated.go	2020-03-04 04:26:30.111495230 +0000
+++ /tmp/swagger-docs.IZkoOW/src/k8s.io/kubernetes/./staging/src/k8s.io/api/core/v1/types_swagger_doc_generated.go	2020-03-04 05:09:42.551694271 +0000
@@ -1300,7 +1300,7 @@ var map_PersistentVolumeClaimSpec = map[
 	"resources":        "Resources represents the minimum resources the volume should have. More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes#resources",
 	"volumeName":       "VolumeName is the binding reference to the PersistentVolume backing this claim.",
 	"storageClassName": "Name of the StorageClass required by the claim. More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes#class-1",
-	"volumeMode":       "volumeMode defines what type of volume is required by the claim. Value of Filesystem is implied when not included in claim spec. This is a beta feature.",
+	"volumeMode":       "volumeMode defines what type of volume is required by the claim. Value of Filesystem is implied when not included in claim spec.",
```

I think `hack/update-generated-swagger-docs.sh` is the right tool to fix this.

j-griffith commented:
> almost there, but generated doc is out of sync:
> 
> ```
> Generated swagger type documentation is out of date:
> --- /home/prow/go/src/k8s.io/kubernetes/./staging/src/k8s.io/api/core/v1/types_swagger_doc_generated.go	2020-03-04 04:26:30.111495230 +0000
> +++ /tmp/swagger-docs.IZkoOW/src/k8s.io/kubernetes/./staging/src/k8s.io/api/core/v1/types_swagger_doc_generated.go	2020-03-04 05:09:42.551694271 +0000
> @@ -1300,7 +1300,7 @@ var map_PersistentVolumeClaimSpec = map[
>  	"resources":        "Resources represents the minimum resources the volume should have. More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes#resources",
>  	"volumeName":       "VolumeName is the binding reference to the PersistentVolume backing this claim.",
>  	"storageClassName": "Name of the StorageClass required by the claim. More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes#class-1",
> -	"volumeMode":       "volumeMode defines what type of volume is required by the claim. Value of Filesystem is implied when not included in claim spec. This is a beta feature.",
> +	"volumeMode":       "volumeMode defines what type of volume is required by the claim. Value of Filesystem is implied when not included in claim spec.",
> ```
> 
> I think `hack/update-generated-swagger-docs.sh` is the right tool to fix this.

It is, and tha'ts what I used, then the rebase happened.

j-griffith commented:
/test pull-kubernetes-e2e-kind

j-griffith commented:
/retest

jsafrane commented on pkg/api/persistentvolumeclaim/util_test.go:
```diff
@@ -167,36 +167,15 @@ func TestPVCDataSourceSpecFilter(t *testing.T) {
 		},
-		"disabled with invalid spec": {
-			spec:        invalidSpec,
-			gateEnabled: false,
```
I should have spotted it earlier: `gateEnabled` is now useless and should be removed from the struct.

smarterclayton commented on staging/src/k8s.io/api/core/v1/generated.proto:
```diff
@@ -2637,11 +2637,14 @@ message PersistentVolumeClaimSpec {
-  // If the provisioner does not support VolumeSnapshot data source, volume will
+  // This field can be used to specify either:
+  // * An existing VolumeSnapshot object (snapshot.storage.k8s.io/VolumeSnapshot - Beta)
+  // * An existing PVC (PersistentVolumeClaim - GA)
```
You shouldn't have to say PVC is GA (since anything not called out as alpha or beta is implicitly GA).  Remove the `- GA` part (but keep `- Beta` on snapshot).

smarterclayton commented:
One more tweak and I will approve this.

j-griffith commented:
> Also, the "Fixes" bug in the PR description looks wrong, should that be pointing to an enhancement?

Yep, so the macro condensed that to just the issue number which your issue in k8s is the same number as my kep in the enhancements repo.  I'll figure out how to address that as well, thanks!

smarterclayton commented:
/approve

jsafrane commented:
/lgtm

j-griffith commented:
/test pull-kubernetes-e2e-gci-gce-autoscaling
/test pull-kubernetes-e2e-kind

saad-ali commented:
/priority important-soon
